did-ion/src/sidetree.rs

```diff
@@ -0,0 +1,1926 @@
+                };
+                DIDStatePatch::AddServices {
+                    services: vec![entry],
+                }
+            }
+            DIDDocumentOperation::RemoveVerificationMethod(did_url) => {
+                let id = did_url.to_string();
+                DIDStatePatch::RemovePublicKeys { ids: vec![id] }
+            }
+            DIDDocumentOperation::RemoveService(did_url) => {
+                let id = did_url.to_string();
+                DIDStatePatch::RemoveServices { ids: vec![id] }
+            }
+        })
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct SidetreeError {
+    // List of error codes: https://github.com/decentralized-identity/sidetree/blob/v1.0.0/lib/core/versions/1.0/ErrorCode.ts
+    pub code: String,
+    pub message: Option<String>,
+}
+
+impl fmt::Display for SidetreeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Sidetree error {}", self.code)?;
+        if let Some(ref message) = self.message {
+            write!(f, ": {}", message)?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl<S: Sidetree + Send + Sync> DIDMethod for SidetreeClient<S> {
+    fn name(&self) -> &'static str {
+        S::METHOD
+    }
+
+    fn to_resolver(&self) -> &dyn DIDResolver {
+        match self.resolver {
+            Some(ref res) => res,
+            None => &NoOpResolver,
+        }
+    }
+
+    fn create(&self, create: DIDCreate) -> Result<DIDMethodTransaction> {
+        let DIDCreate {
+            recovery_key,
+            update_key,
+            verification_key,
+            options,
+        } = create;
+        ensure!(is_empty(&options), "Create options not supported");
+        let (update_pk, recovery_pk, patches) =
+            new_did_state::<S>(update_key, recovery_key, verification_key)
+                .context("Prepare keys for DID creation")?;
+        let operation = S::create_existing(&update_pk, &recovery_pk, patches)
+            .context("Construct Create operation")?;
+        Self::op_to_transaction(operation).context("Construct create transaction")
+    }
+
+    /// <https://identity.foundation/sidetree/api/#sidetree-operations>
+    async fn submit_transaction(&self, tx: DIDMethodTransaction) -> Result<Value> {
+        let op = Self::op_from_transaction(tx)
+            .context("Convert DID method transaction to Sidetree operation")?;
+        let endpoint = self
+            .endpoint
+            .as_ref()
+            .ok_or(anyhow!("Missing Sidetree REST API endpoint"))?;
+        let url = format!("{}operations/", endpoint);
+        let client = Client::builder().build().context("Build HTTP client")?;
+        let resp = client
+            .post(url)
+            .json(&op)
+            .header("Accept", "application/json")
+            .header("User-Agent", ssi::USER_AGENT)
```

Comment:
Could maybe share the client in the struct, and maybe have a helper in `ssi` to make sure we always have this header?

Reply:
Sharing the client would be good to allow connection reuse.
I think the User-Agent header should include the did-ion crate version also. (similar for the other DID method crates)